guillotina/traversal.py

```diff
@@ -286,6 +286,7 @@ def __init__(self, resource, request, view):
                 view_result = await self.view()
                 await commit(warn=False)
             except (ConflictError, TIDConflictError):
+                await abort()
```

Comment:
The tpc handling in `transaction.py` already does some abort logic. Are we sure this is necessary here?

Reply:
Here is the traceback:
```
2019-10-11T11:52:42.996252377Z Traceback (most recent call last):
2019-10-11T11:52:42.996257215Z   File "/usr/local/lib/python3.7/site-packages/guillotina/factory/app.py", line 151, in _handle
2019-10-11T11:52:42.996261229Z     return await super()._handle(request)
2019-10-11T11:52:42.996264936Z   File "/app/api/utils/sentry.py", line 53, in sentry_app_handle
2019-10-11T11:52:42.996268793Z     return await inner()
2019-10-11T11:52:42.99627238Z   File "/app/api/utils/sentry.py", line 49, in inner
2019-10-11T11:52:42.996276471Z     reraise(*_capture_exception(hub))
2019-10-11T11:52:42.996280525Z   File "/usr/local/lib/python3.7/site-packages/sentry_sdk/_compat.py", line 57, in reraise
2019-10-11T11:52:42.996284362Z     raise value
2019-10-11T11:52:42.996287825Z   File "/app/api/utils/sentry.py", line 45, in inner
2019-10-11T11:52:42.996291375Z     response = await old_handle(self, request)
2019-10-11T11:52:42.996295131Z   File "/usr/local/lib/python3.7/site-packages/aiohttp/web_app.py", line 458, in _handle
2019-10-11T11:52:42.996298855Z     resp = await handler(request)
2019-10-11T11:52:42.996302468Z   File "/usr/local/lib/python3.7/site-packages/aiohttp/web_middlewares.py", line 119, in impl
2019-10-11T11:52:42.996306198Z     return await handler(request)
2019-10-11T11:52:42.996309825Z   File "/usr/local/lib/python3.7/site-packages/guillotina_prometheus/middleware.py", line 25, in __call__
2019-10-11T11:52:42.996325607Z     resp = await self.handler(request)
2019-10-11T11:52:42.996329334Z   File "/usr/local/lib/python3.7/site-packages/guillotina/traversal.py", line 286, in handler
2019-10-11T11:52:42.99633304Z     view_result = await self.view()
2019-10-11T11:52:42.996336637Z   File "/usr/local/lib/python3.7/site-packages/guillotina/api/service.py", line 150, in _call_original
2019-10-11T11:52:42.996340332Z     return await self.__original__()
2019-10-11T11:52:42.996343808Z   File "/app/api/services/upsert.py", line 19, in __call__
***
2019-10-11T11:52:42.996347471Z     await get_transaction().commit()
***
2019-10-11T11:52:42.996350967Z   File "/usr/local/lib/python3.7/site-packages/guillotina/db/transaction.py", line 308, in commit
2019-10-11T11:52:42.996354716Z     return await self._commit()
2019-10-11T11:52:42.996358213Z   File "/usr/local/lib/python3.7/site-packages/guillotina/db/transaction.py", line 324, in _commit
2019-10-11T11:52:42.996361855Z     await self.tpc_commit()
2019-10-11T11:52:42.996365698Z   File "/usr/local/lib/python3.7/site-packages/guillotina/db/transaction.py", line 380, in tpc_commit
2019-10-11T11:52:42.996369382Z     await self._store_object(obj, oid)
2019-10-11T11:52:42.996372906Z   File "/usr/local/lib/python3.7/site-packages/guillotina/db/transaction.py", line 366, in _store_object
2019-10-11T11:52:42.996376741Z     await self._manager._storage.store(uid, serial, writer, obj, self)
2019-10-11T11:52:42.996380556Z   File "/usr/local/lib/python3.7/site-packages/guillotina/db/storages/pg.py", line 904, in store
2019-10-11T11:52:42.996384249Z     writer,
2019-10-11T11:52:42.996388484Z guillotina.exceptions.TIDConflictError: Mismatch of tid of object being updated. This is likely caused by a cache invalidation race condition and should be an edge case. This should resolve on request retry..
```

Looking closely to the traceback I see that I'm calling (in a custom service) `get_transaction().commit()` and now I see that the logic to abort a commit is in the transaction manager. If i'd rewrite to `transaction_manager.commit(txn)` this would solve the problem. 

Today I learned that `txn.commit()` is not the same as `tm.commit(txn)` :P

Reply:
ouch yes, we really never want to call `.commit()` directly on a transaction object usually...

Reply:
And is there any reason to put the logic to close the connection in the tm instead of the txn itself? it looks the same to me...

Reply:
I think we could probably refactor tm out of guillotina but it would be quite a change.

Right now, tm is responsible for managing transactions and the connections associated with it.

I think it's not a great transient dependency management we put on tm to do it this way with the connection. But alas, we live with the poor decisions we've made and have to be careful about changing them later.

I would be open to it for G6!

Reply:
:+1: